source/agora/common/TransactionPool.d

```diff
@@ -0,0 +1,226 @@
+
+    Contains a transaction pool that is serializable to disk,
+    using SQLite as a backing store.
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.TransactionPool;
+
+import agora.common.Data;
+import agora.common.Hash;
+import agora.common.Serializer;
+import agora.common.Transaction;
+
+import d2sqlite3.database;
+import d2sqlite3.library;
+import d2sqlite3.results;
+import d2sqlite3.sqlite3;
+
+import vibe.core.log;
+
+import std.algorithm;
+import std.exception : enforce;
+import std.file : exists;
+import std.range;
+
+/// Ditto
+public class TransactionPool
+{
+    /// init sqlite
+    shared static this ()
+    {
+        .initialize();
+    }
+
+    /// deinit sqlite
+    shared static ~this ()
+    {
+        .shutdown();
+    }
+
+    /// SQLite db instance
+    private Database db;
+
+
+    /***************************************************************************
+
+        Params:
+            db_path = path to the database file, or in-memory storage if
+                      :memory: was passed
+
+    ***************************************************************************/
+
+    public this (in string db_path)
+    {
+        const db_exists = db_path.exists;
+        if (db_exists)
+            logInfo("Loading database from: %s", db_path);
+
+        // note: can fail. we may want to just recover txes from the network instead.
+        this.db = Database(db_path);
+
+        if (db_exists)
+            logInfo("Loaded database from: %s", db_path);
+
+        // create the table if it doesn't exist yet
+        this.db.execute("CREATE TABLE IF NOT EXISTS tx_pool " ~
+            "(key BLOB PRIMARY KEY, val BLOB NOT NULL)");
+    }
+
+    /***************************************************************************
+
+        Shut down the database
+
+        Note: this method must be called explicitly, and not inside of
+        a destructor.
+
+    ***************************************************************************/
+
+    public void shutdown ()
+    {
+        this.db.close();
+    }
+
+    /***************************************************************************
+
+        Add a transaction to the pool
+
+        Params:
+            tx = the transaction to add
+
+        Throws:
+            SqliteException if the transaction failed to be added
+
+    ***************************************************************************/
+
+    public void add (Transaction tx) @trusted
+    {
+        import core.memory;
+        GC.addRoot(tx.inputs.ptr);
+        GC.addRoot(tx.outputs.ptr);
+        db.execute("INSERT INTO tx_pool (key, val) VALUES (:k, :v)",
+            hashFull(tx)[], (cast(ubyte*)&tx)[0 .. Transaction.sizeof]);
```

Comment:
This was a start, but it can't work this way. We can't store pointers to disk.

We should probably incorporate (some form) of Contiguous support, and store that serialization here.

Also, SQLite supports memory-mapping through a pragma. I'm looking into that.